File: foliage-rpc/src/main/java/io/foliage/netty/rpc/core/NamedThreadFactory.java
package io.foliage.netty.rpc.core;

import java.util.concurrent.ThreadFactory;
import java.util.concurrent.atomic.AtomicInteger;

public class NamedThreadFactory implements ThreadFactory {

    private static final AtomicInteger threadNumber = new AtomicInteger(1);

    private final AtomicInteger mThreadNum = new AtomicInteger(1);
    private final String prefix;
    private final boolean daemonThread;
    private final ThreadGroup threadGroup;

    public NamedThreadFactory() {
        this("rpcserver-threadpool-" + threadNumber.getAndIncrement(), false);
    }

    public NamedThreadFactory(String prefix) {
        this(prefix, false);
    }

    public NamedThreadFactory(String prefix, boolean daemon) {
        this.prefix = prefix + "-thread-";
        daemonThread = daemon;
        SecurityManager securityManager = System.getSecurityManager();
        threadGroup = securityManager == null ? Thread.currentThread().getThreadGroup()
                : securityManager.getThreadGroup();
    }

    public Thread newThread(Runnable r) {
        String name = prefix + mThreadNum.getAndIncrement();
        Thread t = new Thread(threadGroup, r, name, 0);
        t.setDaemon(daemonThread);
        return t;
    }

    public ThreadGroup getThreadGroup() {
        return threadGroup;
    }
}

File: foliage-rpc/src/main/kotlin/io/foliage/netty/rpc/core/NamedThreadFactory.kt
package io.foliage.netty.rpc.core

import java.util.concurrent.ThreadFactory
import java.util.concurrent.atomic.AtomicInteger

class NamedThreadFactory : ThreadFactory {

    private val threadPoolNum = AtomicInteger(1)
    private val threadNum = AtomicInteger(1)
    private val threadGroup: ThreadGroup
    private val prefix: String
    private val daemon: Boolean

    constructor() {
        this.prefix = "rpcserver-threadpool-" + threadPoolNum.andIncrement
        this.daemon = false
    }

    constructor(prefix: String) {
        this.prefix = prefix
        this.daemon = false
    }

    constructor(prefix: String, daemon: Boolean) {
        this.prefix = prefix + "-thread-"
        this.daemon = daemon
    }

    init {
        val security = System.getSecurityManager()
        this.threadGroup = if (security == null) Thread.currentThread().threadGroup else security.threadGroup
    }

    override fun newThread(r: Runnable): Thread {
        val t = Thread(threadGroup, r, prefix + threadNum.andIncrement, 0)
        t.isDaemon = daemon

        return t
    }
}